tmail-backend/webadmin/webadmin-mailbox/src/main/java/com/linagora/tmail/webadmin/cleanup/CleanupTrashService.java

```diff
@@ -0,0 +1,100 @@
+package com.linagora.tmail.webadmin.cleanup;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.james.core.Username;
+import org.apache.james.mailbox.DefaultMailboxes;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.model.FetchGroup;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.model.MessageRange;
+import org.apache.james.mailbox.model.MessageResult;
+import org.apache.james.mailbox.store.StoreMailboxManager;
+import org.apache.james.task.Task;
+import org.apache.james.user.api.UsersRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.linagora.tmail.james.jmap.settings.JmapSettings;
+import com.linagora.tmail.james.jmap.settings.JmapSettingsRepository;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import scala.collection.JavaConverters;
+
+public class CleanupTrashService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CleanupTrashService.class);
+    public static final String WEEKLY_PERIOD = "weekly";
+    public static final String MONTHLY_PERIOD = "monthly";
+    public static final String TRASH_CLEANUP_ENABLED_SETTING = "trash.cleanup.enabled";
+    public static final String TRASH_CLEANUP_PERIOD_SETTING = "trash.cleanup.period";
+    private final UsersRepository usersRepository;
+    private final JmapSettingsRepository jmapSettingsRepository;
+    private final StoreMailboxManager storeMailboxManager;
+    private final Clock clock;
+
+    @Inject
+    public CleanupTrashService(UsersRepository usersRepository, JmapSettingsRepository jmapSettingsRepository, StoreMailboxManager storeMailboxManager, Clock clock) {
+        this.usersRepository = usersRepository;
+        this.jmapSettingsRepository = jmapSettingsRepository;
+        this.storeMailboxManager = storeMailboxManager;
+        this.clock = clock;
+    }
+
+    public Mono<Task.Result> cleanupTrash() {
+        return Flux.from(usersRepository.listReactive())
+            .flatMap(this::checkSettingAndCleanupTrashForSingleUser)
+            .reduce(Task.Result.COMPLETED, Task::combine);
+    }
+
+    private Mono<Task.Result> checkSettingAndCleanupTrashForSingleUser(Username username) {
+        return Mono.from(jmapSettingsRepository.get(username))
+            .flatMap(jmapSettings -> getJMAPSettingsValue(jmapSettings, TRASH_CLEANUP_ENABLED_SETTING)
+                .map(trashCleanupEnabled -> checkSetting(trashCleanupEnabled, cleanupTrashForSingleUser(username,
+                    getJMAPSettingsValue(jmapSettings, TRASH_CLEANUP_PERIOD_SETTING).orElse(new JmapSettings.JmapSettingsValue(MONTHLY_PERIOD)).value())))
+                .orElse(Mono.just(Task.Result.COMPLETED)))
```

Comment:
Use `filterWhen` + `flatMap` + `switchIfEmpty`

And method extraction

Reply:
fixed in checkSetting method